sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala

```diff
@@ -318,10 +318,33 @@ private[parquet] class ParquetRowConverter(
         new ParquetMapConverter(parquetType.asGroupType(), t, updater)
 
       case t: StructType =>
+        val wrappedUpdater = {
+          // SPARK-30338: avoid unnecessary InternalRow copying for nested structs:
+          if (updater.isInstanceOf[RowUpdater]) {
+            // `updater` is a RowUpdater, implying that the parent container is a struct.
+            // We do NOT need to perform defensive copying here because either:
+            //
+            //   1. The path from the schema root to this field consists only of nested
```

Comment:
When we have deeply nested struct inside an array, is it the first case here?

I think it is fine because at the element converter the top level struct inside an array element will do the defensive copying. So in nested struct converter, we will see RowUpdater from parent struct so don't need defensive copying too.

Just maybe good to also update it in the doc.

Reply:
Yes, that's right. After thinking about this some more, I think I've come up with a clearer explanation and have updated the code comment: https://github.com/apache/spark/pull/26993/commits/4651b2fd724a56515c087903284682c9ba947c31